### ISSUE TITLE
Add timeout option

### ISSUE THREAD

**Issue by wheresrhys** [whatwg/fetch#20]
Most ajax libraries and xhr itself offer a convenient way to set a timeout for a request. Adding `timeout` to the properties accepted in the options object therefore seems like a sensible addition to fetch.

https://github.com/github/fetch/pull/68

**Comment by annevk** [comment]
Paging @domenic. We could add this but it seems like we should have a sensible termination strategy first upon which this could be layered.

**Comment by domenic** [comment]
Ideally we would have a design for cancelable promises that fetch could use. But unfortunately there are still four or five competing proposals thrown out and every time a web developer asks me "why can't we do cancelable promises?" and I look into exactly what they're asking for I end up adding a new proposal to that list. It would take some concerted effort to try to find something good.

In the meantime I think a sensible cancelation strategy is to reject the promise with some sort of error. The actual cancel() method if it exists should not be on the returned promise though, at least not until we figure out cancelable promises. That leaves a couple options I can see:

- Don't let people cancel requests until the headers come back and you have a response object. In this case you could use the built in stream cancelation of ReadableStream: `res.body.cancel()`.
- Come up with something more awkward. E.g., `fetch.cancelable(...)` returns a `{ promise, cancel }` tuple. Not horrible with destructuring but pretty awkward. And will likely be obsoleted by cancelable promises.

As for the error itself we could future-proof there slightly better. If we create a new `Error` subclass called e.g. `Cancellation` that follows all the normal ES conventions for Error subclasses, in the worst case it ends up being fetch-only but in the best case ES ends up using it for cancellable promises.

**Comment by matthew-andrews** [comment]
> Don't let people cancel requests until the headers come back and you have a response object. In this case you could use the built in stream cancelation of ReadableStream: res.body.cancel().
This doesn't feel particularly useful… the cases where you would want a fetch to timeout would be when the network conditions were not good and you weren't getting anything back…

Given timeout is such a common use case, even if promises had a concept of being cancelled providing `timeout` option seems immensely useful… what am I not seeing here?

**Comment by pornel** [comment]
To me it seems that having API to cancel started fetch is completely separate from request timeouts. Mechanism used for cancellation on timeout can be completely private (for now).

**Comment by annevk** [comment]
@pornel yes, perhaps. If we expose termination I would expect it to be on `Request` myself, as we also plan to expose `Request` elsewhere, on elements and such.

**Comment by annevk** [comment]
@matthew-andrews baby steps. It'll come.

**Comment by wheresrhys** [comment]
> Ideally we would have a design for cancelable promises that fetch could use. 

What are the reasons that rejection of an ordinary promise would be an unacceptable solution?@domenic

**Comment by annevk** [comment]
How do you envision that would work?

**Comment by wheresrhys** [comment]
Assuming that fetch creates the promise, then can't it just have a setTimeout* within the executor function which calls reject? Why is that not feasible?

*or equivalent in whichever language the browser implements it in

**Comment by wanderview** [comment]
If this is spec'd, please also define behavior for timeouts occurring after headers are received, but before the end of the content body.

**Comment by Jxck** [comment]
same question here, I wanna abort() request.

I think what we need is `abort()`. timeout is enable using setTimeout + abort.
and in the fetch spec, I think the problem is fetch doesn't has a instance, only a function.
fetching process needs a internal state, and will resolve at endpoint state, reject otherwise.

in my opinion will like this.

```js
var req = new Request(url, option);
req.fetch().then((res) => {
  console.log(res.status);
}).catch((err) => {
  console.error(err); // this will also happen in `abort()` request
});

// timeout a request
setTimeout(function() {
  req.abort(); // reject the fetching process
}, 1000);
```

**Comment by annevk** [comment]
I agree that we want something like that. It's a bit unclear to me still what the exact semantics should be.

**Comment by matthew-andrews** [comment]
As one possible option, `node-fetch` has implemented `timeout` — https://github.com/bitinn/node-fetch#options — might be worth looking at/considering…  it's simple :smile:

**Comment by annevk** [comment]
The semantics are still unclear. Does the timeout run until the promise returns? Does the timeout terminate the underlying stream somehow once the promise has returned?

**Comment by Jxck** [comment]
@matthew-andrews I think timeout should be Hi-level API on `abort()`.

@annevk I think node.js http module will help us including make res as streams for getting progress event.
http://nodejs.org/api/http.html#http_http_request_options_callback

**Comment by pornel** [comment]
There's an interesting difference between "deadline" and "timeout".

From [`man ping`](http://linux.die.net/man/8/ping): 

* `-w deadline`
          Specify a timeout, in seconds, before ping exits regardless of how many packets have been sent or received. In this case ping does not stop after count packet are sent, it waits either for deadline expire or until count probes are answered or for some error
          notification from network.

* `-W timeout`
          Time to wait for a response, in seconds. The option affects only timeout in absense of any responses, otherwise ping waits for two RTTs.


The "deadline" option is implementable using `setTimeout()`, but an actual timeout, when measured as time from the last packet from the server, is actually more interesting and useful, and something that only browser can implement. I'd definitely like timeout to mean the latter, because it can distinguish between connection that's dead and a working connection that's just slow.

**Comment by Jxck** [comment]
that's interesting but is that scope of fetching ?
network programming has tons of configurable options,
but adding much of them will make spec like a Camel.

in fetch context, I think `abort` should same behave on XHR,
because XHR could build upon fetch api.

in application, that will use for terminate the process gracefully, 
and getting back flow control to user land javascript.
(ex, force reject promise, force call a callback)

If you wanna do something like that, raw-socket API may works.

**Comment by pornel** [comment]
@Jxck Yes, I think timeout is in scope of fetching, and that's an important feature.

In the [FT app](http://app.ft.com) we've had to add timeouts to every network request, because we learned that on some networks requests just never finish, and that can make the app unusable. Now we're treating network requests without a timeout as a bug.

I'm not opposed to having an API to abort requests. I think these are largely independent problems: it's possible to have timeouts without a public abort API, and it's possible to have only abort API and leave timeout implementation to the user. I think having both would be the best.

If the decision was to _only_ have an abort API instead of support for timeout, then that would be workable, but not ideal — I'd always have to wrap fetch in `setTimeout` myself, and perhaps add even more elaborate wrapper if I wanted to implement timeout from the last received bit of data, rather than a simple "deadline" option that could unnecessarily throw away data on slow connections.

**Comment by Jxck** [comment]
ah ok you talking about `timeout option`, not `timeout()` method.
that seems to add option to `RequestInit` https://fetch.spec.whatwg.org/#requestinit
(add or not is another discussion)

adding a option doesn't breaking changes of `fetch(url, option)`.
but adding a programmable `abort()` or `progress` event may change interface.

in some browser start to implements fetch(), so I think we need to start discussion about
breaking changes immediately if we do. (or too late to change fetch api already?)

**Comment by annevk** [comment]
The problem with `Request.prototype.abort()` is that at the moment `fetch()` clones its input. That is, the current design is such that you can `fetch()` a single `Request` instance multiple times (unless it has a body).

(There is also some complication with `Request` being used as cache key and as a way to tell a service worker what a client (document/worker) wants.)

Would need to think a bit about how we can change some of these characteristics to make it all fit better.

**Comment by wojciak** [comment]
It depends how high of an API fetch should be. If it's a low-level request - we don't need abort. We'll wrap it in a higher level promise API and that's it.

If on the other hand it's high-level and we want to abort, it should be instantiated and work like a resource so: var someResource = new Fetch(); someResource.abort(); 

The first option imho is better.

**Comment by annevk** [comment]
Discussion for `abort()` moved to https://github.com/slightlyoff/ServiceWorker/issues/625 with a subclassed promise returned from `fetch()` being the leading candidate to expose it.

Leaving this open for the timeout idea from @pornel which seems like a rather interesting primitive to expose, network stacks allowing.

**Comment by Jxck** [comment]
doing fetch with just one function is difficult to expand with adding api in future.
(anti Open-Closed Principle)

in current interface, we do all thing with 
- adding new props to option arguments
- adding new methods to returned promise

the problem is current `fetch()` is a high level api.
but I believe we need a low level api for fetching
what has capability for building XHR or kind of high-level api on top of it.

I heard that chrome are going to export fetch to window in M42 (mid April)
If that happens, it became hard to change fetch interface.
I think we need to discuss more about current api while we have a chance to change api.

**Comment by annevk** [comment]
What is wrong with `fetch()` as a low-level API? It seems to me that the amount of extensibility hooks we have is fine. We have the `Request` object for initialization, and at some point we'll have the `FetchPromise` object for status and control. What else is needed?

**Comment by annevk** [comment]
Also, I would appreciate it if you could move such a discussion to a new issue. Or perhaps we should move timeout to a new issue as this is getting rather cluttered.

**Comment by monolithed** [comment]
Ho to use the fetch without aborting?

**Comment by Mouvedia** [comment]
You need an xhr-parity label on this issue.

For the ones confused, we are requesting `XMLHttpRequest.timeout` which is either an unsigned long (Mozilla) or Integer (Microsoft) representing millisecond(s).

**Comment by danostrowski** [comment]
As @wheresrhys and @pornel have pointed out, it seems *very* trivial to pass along a timeout option and add an `ontimeout` handler that calls reject in the construction of the Promise returned by `fetch`.

For people moving from jQuery.ajax and the like, this is a barrier that should be easy to fix. Should we just submit a patch?

**Comment by annevk** [comment]
It's not actually that easy, given CORS preflights, HTTP authentication, involvement of service workers, and the choice mentioned in https://github.com/whatwg/fetch/issues/20#issuecomment-73727547. Also, if we get abort semantics, we might want to use those here and not reject, so that is unclear too at this point.

**Comment by danostrowski** [comment]
Hmmm. Interesting.

XMLHttpRequests work with CORS preflights and HTTP authentication, right? Why would this be any different?

I confess I'm not familiar with how "involvement of service workers" affects this, is there some comment that explains what negative interaction might happen with timeout? The mechanism for handling `.ontimeout` seems the same as `.onload` in the context of the main `.fetch` operation.

Regarding comment #20, is there a compelling reason to make timeout behave any differently from how it's already defined on XHR? This isn't ping; seems like a non sequitur.

> Also, if we get abort semantics, we might want to use those here and not reject, so that is unclear too at this point.

You mean... change the object you get out of `.fetch` to not be a standard JavaScript Promise?

**Comment by Krinkle** [comment]
@danostrowski How to abort a request is relatively straight forward. (Or at least outside the scope of this issue).

Unfortunately, it seems we're blocked on figuring out a way to expose this feature. E.g. do we add an `.abort()` method to the Promise returned by `fetch()`? And how does this interact with other promises after calling `.then()`. See whatwg/fetch#27.

**Comment by danostrowski** [comment]
I wasn't proposing an abort, I don't see it as a necessity, from the
outside. A timeout argument to xhr suffices which simply results in .reject
being called which, in turn, works automatically with the Promise
methodology.

**Comment by pornel** [comment]
@Krinkle I don't want `abort()` or any other API to abort requests externally. What I need is rejection of the promise if the network is unusable or stalled. This is a different thing.

**Comment by danostrowski** [comment]
@pornel the problem I see with what you wrote is that the `xhr` object in `oldFetch` will NOT have a timeout so it might complete and call `resolve` itself or, also confusing, fail and you'd get two `reject` calls. Am I reading that right? What would stop that?

I think you *have* to set `xhr.timeout = X;` in fetch and then use `xhr.ontimeout` to call `reject` because ultimately the `xhr` object is the source of truth about whether to proceed or not.

Basically you just need to add to `fetch`:

```javascript
xhr.timeout = request.options.timeout || 0;
```
sometime before the `.send` and something like this near where onerror is defined:

```javascript
xhr.ontimeout = function() {
    reject(new TypeError('Request timeout'));
}
```

then this should work as expected...

```javascript
fetch('/users', {
  timeout: 300
}).then(function(response) {console.log('made it!');
}).catch(function(e) {console.log(e);})
```

**Comment by jonathan-fielding** [comment]
@pornel I agree with your example, we need to be able to pass a timeout as part of the options and then the error passed to the rejected promise should tell us that the timeout was exceeded.

Would also be useful to be able to timeout, do something to handle timeout condition and then if the response eventually comes back to be able to resume  (kinda like when you regain a connection perhaps). Not sure if this is asking a bit to much from fetch API though.

**Comment by peteruithoven** [comment]
Any progress on this?

**Comment by anka-213** [comment]
What is blocking this issue? As far as I can see, this is not dependent on #27 (which is obviously not going to be solved in the near future). All that should be needed is an option for timeout and to reject the promise on timeout. That doesn't seem controversial.

**Comment by DiThi** [comment]
I subscribed to this issue time ago because I need connections to fail when they're stuck. However, it's common to have many requests pending because the browser only downloads a limited number simultaneously. Therefore we need a way to timeout a whole group of promises when there hasn't been activity for X amount of time.

I've been using the `timeout` property of XHR but it turns out it's a timeout since `send()` was called, which is like setting a `setTimeout` regardless of traffic. I had to remove the `timeout` because it always failed on slow connections.

In other words, we need several things, independently:
* XHR style timeout, a.k.a. **deadline**: reject the promise when it took too long. Seems easy to polyfill.
* Abort: not needed for this specific issue but very useful (e.g. to cut in traffic and leave room for an alternative, smaller request). #27
* **Timeout since last received data. IMHO, the spec should specify the client will handle stuck files and re-start the connection.**

**Comment by annevk** [comment]
@DiThi note that XMLHttpRequest does offer that ability since you can keep increasing the timeout as data arrives. Whatever we do here though does depend on #27 and corresponding questions for streams, since they all interact.

**Comment by pornel** [comment]
> Whatever we do here though does depend on #27

No, it does not. "Provide developers a method to abort" is very different than "implementation can internally abort and reject the promise without giving developers any other API to abort". 

What about https://github.com/whatwg/fetch/issues/20#issuecomment-135426968 ?

**Comment by annevk** [comment]
So you only want the timeout to take effect for the period of time until all HTTP headers have been received? Any network slowness after that is not important?

**Comment by pornel** [comment]
Yes, for an "MVP" that would be useful already, as in my experience most stuck connections happen before server response.

**Comment by pornel** [comment]
Ideally I'd like the timeout to apply to streams too (and be counted since last data received), but I'd rather have something sooner, rather than wait another year debating another API that I wouldn't use.

In the codebase I'm working on right now I've got 33 uses of request timeout, 0 uses of request abort.

**Comment by DiThi** [comment]
@annevk I did exactly that increasing the timeout after an error. But imagine a file that takes more than a minute to be downloaded in a slow connection, with a minute of timeout. It ends taking twice as much time and bandwith! (even more when the timeout was half a minute).

More than **abort**, we need **progress**. Having both we can detect stuck connections and restart them. Having just abort we can easily emulate the XHR timeout feature (to avoid downloading the same data several times).

In other words: We need first a way to measure download progress. Then a way to abort the connection. Not necessarily exposed to JS, but just usable by the **timeout** and **deadline** implementations.

Maybe we should split the issue in two? Timeout for stuck connections (not just slow ones) and deadline (if it's not downloaded in X time, I don't need it; a.k.a. XHR timeout).

**Comment by jakearchibald** [comment]
We should offer lower-level APIs so more combinations can be composed.

We already have a time api - `setTimeout`. Streams will give you insight into download progress (although we may want to consider how uploads are observed, especially if redirects happen). Abort is the other missing piece.

**Comment by pornel** [comment]
Yes, I think splitting the issue into simple deadline and more advanced use cases would help.

**Comment by alcat2008** [comment]
I think the code below would be helpful.

```
var p = Promise.race([
  fetch('/resource-that-may-take-a-while'),
  new Promise(function (resolve, reject) {
    setTimeout(() => reject(new Error('request timeout')), 5000)
  })
])
p.then(response => console.log(response))
p.catch(error => console.log(error))
```

**Comment by DiThi** [comment]
@alcat2008 That code is very useful, thanks.

However it fails to avoid **data waste** when it hasn't started yet (#179) or when it's a big file that is expected to take a while to finish (#180). Both things happened to me frequently over an ADSL connection.

Since two separate issues were made and the solution outside these two is just to use race(), I propose to close this one.

**Comment by buraktamturk** [comment]
Something like this could also allow us to cancel requests manually (without timeout).

```javascript

var cancelpromise1 = new Promise(function(resolve, reject) {
     setTimeout(resolve, 1000);
});

var cancelpromise2 = /* a promise that will be resolved when user clicks cancel */;

var lastpromise = Promise.race([cancelpromise1, cancelpromise2]);

fetch(..., { timeout: lastpromise })
   ...

```

**Comment by CamiloTerevinto** [comment]
It's been well over an year since the last comment, is there any progress on any of this?

**Comment by annevk** [comment]
We're pretty close to fixing #447. That should make it a lot easier to implement this in a library and eventually get something standardized.

**Comment by mattlubner** [comment]
Forgive me if I'm being obtuse, but how does clearing the timeout after the response resolves timeout only the request? If `clearTimeout(timeoutId)` is after `await fetchPromise`, it will only clear the timeout after the response is received. Unless I'm missing something?

**Comment by CamiloTerevinto** [comment]
@annevk thanks, I did not see that issue. 

@jakearchibald is there any timeline for that to be production-ready or are you implying I can use that already?  I also have the same doubt as matlubner.

**Comment by jokeyrhyme** [comment]
@CamiloTerevinto @mattlubner I believe `const response = await fetch(url)` gives you a response object that is not necessarily "done", but provides you with body methods to process content that will eventually be available

```js
const response = await fetch(url)
// if you abort here, then some downloading may have happened, but you can avoid the rest
const text = await response.text()
// if you abort here, then you just downloaded the whole thing, wasteful!
```

**Comment by annevk** [comment]
It gives you a response with headers, but without a response body. A response with headers is the earliest signal we're exposing in the web platform. (Note that it doesn't necessarily mean that the request body has been fully transmitted, though typically that's the case.)

**Comment by jakearchibald** [comment]
@CamiloTerevinto Edge & Firefox are already implementing. Chrome will start shortly. I don't know about Safari. Timelines for production-ready depend on the release schedule of individual browsers.

**Comment by CamiloTerevinto** [comment]
@jakearchibald  this might not be the right place, I'm aware, but I was looking into this for the react-native fetch implementation (which as far as I know, relies on this library).

**Comment by HyperSimon** [comment]
@domenic 因噎废食

**Comment by Angelk90** [comment]
@alcat2008 : Hello, I tried the code you proposed but it does not seem to work.
In a nutshell, the page to which I refer to a redirect, this is the problem.
Can you give me some advice?

**Comment by jasonwr** [comment]
@Angelk90 et al, you might want to have a look at this AjaxService I wrote:

https://github.com/github/fetch/issues/617

IMHO **node-fetch** gives you everything the native Fetch API within ES6 does and it provides some additional features including a **timeout**

**Comment by xgqfrms** [comment]
# `credentials: "include"`

https://developers.google.com/web/updates/2015/03/introduction-to-fetch

![image](https://user-images.githubusercontent.com/7291672/41389183-e6bb2de2-6fc1-11e8-9145-c7eaf6e91da9.png)

**Comment by franklinyu** [comment]
@xgqfrms How on earth is that related to this issue?

**Comment by ImanMh** [comment]
2018! do we have timeout?

**Comment by taralx** [comment]
76% yes: https://caniuse.com/#feat=abortcontroller

**Comment by FranklinYu** [comment]
According to Can I Use link above, Safari doesn't support `AbortController`, but MDN says otherwise: https://developer.mozilla.org//en-US/docs/Web/API/AbortController#Browser_compatibility

How can I verify this? I tried this JavaScript code:

```javascript
const controller = new AbortController()
const signal = controller.signal

document.getElementById('download').addEventListener('click', fetchVideo)
document.getElementById('abort').addEventListener('click', () => {
    controller.abort()
    console.log('Download aborted')
})

// some random large image I found online
const url = 'https://upload.wikimedia.org/wikipedia/commons/6/6e/Monasterio_Khor_Virap%2C_Armenia%2C_2016-10-01%2C_DD_25.jpg'

function fetchVideo() {
    fetch(url, {signal})
        .then( r => r.blob() )
        .then( b => console.log('success') )
        .catch( e => console.error('Download error: ' + e.message) )
}
```

but my Internet is too fast, and the image finished in a second so I cannot hit "Abort" before that. Any other way to test it?

**Comment by jasonwr** [comment]
@FranklinYu the easiest way to verify if it works is to run it in Safari. 💯

**Comment by fastest963** [comment]
@FranklinYu you can use settimeout.io and just request a 5s timeout.

**Comment by styfle** [comment]
@FranklinYu It exists but it doesn't work.

> Safari has window.AbortController defined in the DOM but it's just a stub, it does not abort requests at all. The same issue also affects Chrome on IOS and Firefox on IOS because they use the same WebKit rendering engine as Safari.

See [caniuse](https://caniuse.com/#feat=abortcontroller)

**Comment by jasonwr** [comment]
@FranklinYu I believe that there are tools to slow down your connection with Safari else just go to a local coffee shop.

**Comment by FranklinYu** [comment]
@styfle I saw that note, but I would like to verify this myself before I submit pull request to https://github.com/mdn/browser-compat-data.

@fastest963's suggestion is right on the spot. I'm going to fix the compatibility table.

**Comment by ianstormtaylor** [comment]
Has there been any talk of improving the timeout ergonomics? Now that the whole `AbortController` thing has been hashed out?

It seems like… given the nature of HTTP/TCP, no request is ever guaranteed to respond to you, and the agreed upon solution for this is to use timeouts. Such that pretty much every request you make should have some form of timeout. (Just like every request you make should account for network errors.)

Given that, it seems problematic that it takes…

```js
const controller = new AbortController();
const signal = controller.signal;
const fetchPromise = fetch(url, {signal});
const timeoutId = setTimeout(() => controller.abort(), 5000);
const response = await fetchPromise;
clearTimeout(timeoutId);
```

…to properly send a request with a timeout.

Has anyone discussed adding a `timeout: number` shorthand option? (It was mentioned in the very first comment back in 2015, but it seems like the UX might have gotten lost in the quest for true cancellation?)

```js
fetch(url, {
  timeout: 5000
})
```

**Comment by taralx** [comment]
If you're not making a huge number of requests, you can probably skip the clearTimeout step. Then you can make a `timeoutSignal(ms)` function pretty easily.

**Comment by Pauan** [comment]
@ianstormtaylor I think there's some complicated questions about the timeout.

For example, if you specify `timeout: 5000`, should the timeout also include things like `response.json()`? I imagine most of the time the answer is "yes", but then what if you don't call `response.json()`?

In other words, how does `fetch` know that you're "done", so it can cancel the timer?

So I think this is a better design:

```js
async function with_timeout(ms, f) {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), ms);

    try {
        return await f(controller.signal);

    } finally {
        clearTimeout(timeoutId);
    }
}
```

Now you can use it like this:

```js
with_timeout(5000, (signal) => fetch(url, { signal }))
```

Or if you want it to also cancel the response:

```js
with_timeout(5000, async (signal) => {
    const response = await fetch(url, { signal });
    return await response.json();
})
```

This is a general solution which isn't tied to `fetch`, so you can also use `with_timeout` in other areas, like Streams.

**Comment by ianstormtaylor** [comment]
@Pauan why would you want it to include `response.json()`? I don't think including it gains you anything. 

The way "timeout" is used in most libraries is pretty clear: if you haven't received a response in that amount of time then it's considered failed. And it's defined that way because that's how it's useful to guard against the uncertainty of the network. But if you're about to call `response.json` that means you've already received a response. There's no need to have the timeout extend to that part of the code because there's no longer any uncertainty there?

@Pauan @taralx As for the `withTimeout` helpers, I agree that those _can_ be created. Thanks for the examples! I just think timeouts are such a common case (so much so that arguably you're not writing "complete" code if you aren't including a timeout). And that for common cases like that the API should help you do the right thing, which in this case would be:

```
fetch(url, {
  timeout: 5000
})
```

This seems like at least one thing XHR got right? `xhr.timeout = 500`.

**Comment by ImanMh** [comment]
@ianstormtaylor you got this very correctly. This thread has become the place where some people are trying to convince the others that 'timeout' must be included. I know that I can implement it my self but the entire library can also be implemented.
The point is that this library looks incomplete and unfinished without  timeout functionality. So let's write timeout here instead of expecting everyone to write their own version of it, which is also the whole point of making a library.

**Comment by Bnaya** [comment]
> but the entire library can also be implemented

This is incorrect. also fetch us not a library.
fetch is the most low level http client api that the browser provides. (Even more than XHR)
It first goal is to be powerful primitives without un-handled cases (eg, what do we do with response.body.json()) before it tries to give more high level features, that can be implemented in user land.

**Comment by ImanMh** [comment]
@bnaya you are right, There are a lot of fetch libraries out there that are very much similar to the standards in this repo. My point is that adding timeout to these standards will just improve it and prevents others to write unnecessary extra code.

**Comment by Pauan** [comment]
> @ianstormtaylor But if you're about to call response.json that means you've already received a response. There's no need to have the timeout extend to that part of the code because there's no longer any uncertainty there?

That's not true: when `fetch` resolves that just means that the headers have been received, the body has *not* been received.

So when you fetch the body (using `response.json()` or `response.text()` or similar), it can fail. So you need the timeout to cover that case as well.

> And that for common cases like that the API should help you do the right thing

That's my point: the "right thing" in most cases is that the timeout should extend to retrieving the body. That's very easy to do if you're using `with_timeout`, but tricky to define if `timeout` is native to `fetch`.

> This seems like at least one thing XHR got right?

XHR completes when the *body* is received. `fetch` completes when the *headers* are received. They're not the same thing.

I agree that XHR's behavior is correct, which is why the timeout for `fetch` must include fetching the body. That's what `with_timeout` does.

**Comment by Mouvedia** [comment]
> So when you fetch the body (using response.json() or response.text() or similar), it can fail.

Can you be more precise about this please.
In general it can throw if it's locked, has already been read or canceled. For example:
- `text` check https://stackoverflow.com/a/49786130/248058
- `json` check https://infra.spec.whatwg.org/#parse-json-from-bytes

Where does it say that it can fail because the body wasn't received?
It's a stream, hence it's handled.

**Comment by Pauan** [comment]
@Mouvedia The server might hang (not send a response). That's the whole point of a timeout, to handle the situation where the server doesn't send data.

**Comment by Mouvedia** [comment]
@Pauan Could you point out the type of error (from the spec) that this would produce?

**Comment by Pauan** [comment]
@Mouvedia You are misunderstanding. Failure is not the same as an error. There is no error if the server fails to send data, there is simply nothing.

That is why you are supposed to add in a timeout. The timeout will throw an `AbortError` if the server takes too long to send the data:

https://dom.spec.whatwg.org/#abortcontroller-api-integration

**Comment by Mouvedia** [comment]
I think you are misunderstanding me.

> there is simply nothing.

There is never "nothing". If the "server hangs" and it happens during the streaming, you won't receive another chunk.

Therefor the timeout should be taken into account until the `read`'s reader returns `done` as true.
This cannot be achieved if you are only consuming the body without using `getReader`.

**Comment by Pauan** [comment]
@Mouvedia "Not receiving another chunk" is "nothing". That's the point: you have not received the data, and the server has not sent the data, so you have to sit there with nothing while you wait for the server to send the data.

I'm not sure what you're arguing for. I have always argued that the timeout should include the body (in most situations), which is exactly what my `with_timeout` does.

The problem with making `timeout` native to `fetch` is that you might *not* retrieve the body, you might call `fetch` solely to check the headers. So in that case the `timeout` *shouldn't* take into account the body. So how does `fetch` know whether the timeout should take into account the body or not?

The `with_timeout` function does not have that problem, it makes it easy to choose whether the timeout should include the body or not.

**Comment by Mouvedia** [comment]
As you can read [here](https://xhr.spec.whatwg.org/#the-timeout-attribute), XHR's `timeout` is only taken into account during "fetching". We should match that behaviour to achieve consistency.  

ref https://fetch.spec.whatwg.org/#concept-fetch

**Comment by karlhorky** [comment]
> XHR completes when the body is received. fetch completes when the headers are received. They're not the same thing.

@Pauan Then maybe the timeout for `fetch` should take this into account? Adding one more property to @ianstormtaylor's proposal could be a possible solution, no?

```js
fetch(url, {
  timeout: 5000,
  timeoutIncludesBody: true, // default: false
})
```

Defaulting this to `false` is so that the default behavior of `fetch` (to complete when the headers are received) is respected. This is to provide the least surprising API when one knows how `fetch` works.

I don't think that the solution to "keep this in userland" is great here, because this is a very common use case for non-trivial apps. It should be built into the platform.

**Comment by Mouvedia** [comment]
It's inaccurate or incomplete.

> XHR completes when the body is received. fetch completes when the headers are received.

XHR has `readystate` 3 which corresponds to HEADERS_RECEIVED. We gotta strive for parity with XHR.

ref https://stackoverflow.com/a/40877968/248058

**Comment by jokeyrhyme** [comment]
There's a potentially different kind of timeout here: https://aws.amazon.com/blogs/aws/elb-idle-timeout-control/ 
An "idle timeout", for when there is no activity on a connection for X seconds

e.g. with an idle timeout of 15 seconds:
- a slow connection that trickles out the headers at 10 seconds and a body chunk every 10 seconds will NOT trigger the idle timeout
- a non-streaming server that returns the headers at 1 second, but takes 20 seconds to render the body WILL trigger the idle timeout
- a cellular connection on a train with headers at 1 second, some of the body at 10 seconds, then a 20 second tunnel WILL trigger the idle timeout

I don't think it is always the intention for a timeout to only consider the time until the first byte

I don't think it is always the intention for a timeout to consider full receipt of the headers and body

**Comment by Pauan** [comment]
@Mouvedia As the spec [clearly shows (in step 14)](https://fetch.spec.whatwg.org/#main-fetch), XHR will wait for both the headers *and the body* to complete. The timeout for XHR includes the body. Please read the spec more carefully.

> XHR has readystate 3 which corresponds to HEADERS_RECEIVED. We gotta strive for parity with XHR.

The `load` event will only fire when the body has been received. The timeout cares about the load event, not readystate.

XHR is not the same as `fetch`. XHR *always* retrieves the body, whereas `fetch` lets you choose whether to retrieve the body or not. That's why `fetch` needs to have a choice whether the timeout includes the body or not.

----

@jokeyrhyme That's a good point, that complicates the timeout even more. Which is another argument for having this in user-land, so that way the user has full control, rather than adding in several options to `fetch`.

**Comment by FranklinYu** [comment]
> If the "server hangs" and it happens during the streaming, you won't receive another chunk.

@Mouvedia Yes, you won't receive another chunk, but the promise of `.text()` won't resolve either. It simply wait for the server to recover (possibly up until the network/system timeout which could be 5 minutes or similar).

**Comment by FranklinYu** [comment]
> * a slow connection that trickles out the headers at 10 seconds and a body chunk every 10 seconds will NOT trigger the idle timeout

@jokeyrhyme I doubt whether this can be detected in JavaScript. Now that the body is a stream, is browser allowed to buffer the response? Is browser supposed to trigger any JavaScript event as soon as it receive a single byte from server?

**Comment by ImanMh** [comment]
This thread is open since 2015 and the issue is still present and I don't think this is considered an issue at all, I'm switching to [Axios](https://github.com/axios/axios) it has built it timeout.

**Comment by annevk** [comment]
I noticed some misunderstanding about `fetch()` above. It will retrieve the body, just like XMLHttpRequest. So timeout would work similarly.

We're considering adding timeout support to AbortSignal: https://github.com/whatwg/dom/issues/951. What I'm wondering is whether people need to distinguish between the different reasons a fetch might have aborted or whether that's immaterial.

**Comment by Mouvedia** [comment]
What's annoying is non-parity with XHR: `fetch` will call the signal's abort handler even if the request has been completed—XHR, as expected, doesn't. After the completion the function should become a no-op.

**Comment by annevk** [comment]
That doesn't make sense. You pass a signal to `fetch()` and that will react to the abort signal. If the fetch has completed it'll no-op.

**Comment by Mouvedia** [comment]
`<off topic>`
> If the fetch has completed it'll no-op.

Are you talking about the specification or the implementations? Do you have tests for that?
`</off topic>`

---

Anyway what I meant is, please try to strive for parity with XHR. We don't want more gotchas: the specification needs to be explicit. Ill give you another example, in XHR abortion behaviour on page quit wasn't clear at first.
So first and foremost parity, then you can move on and add new functionalities that are missing.
i.e. don't reactively change how XHR behaves in light of what you are trying to add to fetch (because XHR is _now_ explained in terms of `fetch`)

**Comment by annevk** [comment]
I'm not sure I follow. There are open issues around document unloading, but they affect XMLHttpRequest and `fetch()` equally. And I wasn't even talking about adding anything to fetch here...

**Comment by Mouvedia** [comment]
> And I wasn't even talking about adding anything to fetch here...

From the point of view of developers, adding `timeout` to `AbortSignal` is directly related to XHR obviously.

> We're considering adding timeout support to AbortSignal: whatwg/dom#951.

Are you adding [`timeout` or `deadline`](https://github.com/whatwg/fetch/issues/20#issuecomment-73727547)?

**Comment by annevk** [comment]
It's on top of signal, so deadline, which is also what XMLHttpRequest has. However, we might add signals to reading from a stream, which would allow for the former.

**Comment by Mouvedia** [comment]
Dunno what the others think but it would be better if we could stay consistent regarding terminology.
Personally I don't care what it is called as long as I have at least XHR parity.

**Comment by ririko5834** [comment]
How can I add timeout to request now?

**Comment by saschanaz** [comment]
Should this be considered fixed as we now have `AbortSignal.timeout`?

**Comment by annevk** [comment]
Thanks, yeah, this can be closed. We still have #179 and #180 for the more complicated scenarios.

**Comment by karlhorky** [comment]
`AbortSignal.timeout` looks great, thanks to everyone who worked on speccing this!

```ts
// Fetch the URL, cancelling after 8 seconds
fetch(url, { signal: AbortSignal.timeout(8000) });
```

Apparently [available in Chrome 103 beta](https://blog.chromium.org/2022/05/chrome-103-beta-early-navigation-hints.html#:~:text=abortsignal.timeout()%20static%20method)!!

**Comment by saschanaz** [comment]
And in [Firefox 100](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/timeout#browser_compatibility) 😉

**Comment by karlhorky** [comment]
Thanks, wasn't aware about this, updated my post above!
